### PR TITLE
implement `node:events` in javascript

### LINKED ISSUE #2190: EventEmitter performance issue causes Bun's node:crypto to be 20% slower than node.js's node:crypto
According to this benchmark:

<details>

```ts
const Benchmark = require('benchmark')
const suite = new Benchmark.Suite()
const hash = require('crypto').createHash
const data = 'Delightful remarkably mr on announcing themselves entreaties favourable. About to in so terms voice at. Equal an would is found seems of. The particular friendship one sufficient terminated frequently themselves. It more shed went up is roof if loud case. Delay music in lived noise an. Beyond genius really enough passed is up.'

const cyrb53 = (str: string, seed = 0) => {
  let h1 = 0xdeadbeef ^ seed,
    h2 = 0x41c6ce57 ^ seed;
  for (let i = 0, ch; i < str.length; i++) {
    ch = str.charCodeAt(i);
    h1 = Math.imul(h1 ^ ch, 2654435761);
    h2 = Math.imul(h2 ^ ch, 1597334677);
  }
  
  h1 = Math.imul(h1 ^ (h1 >>> 16), 2246822507) ^ Math.imul(h2 ^ (h2 >>> 13), 3266489909);
  h2 = Math.imul(h2 ^ (h2 >>> 16), 2246822507) ^ Math.imul(h1 ^ (h1 >>> 13), 3266489909);
  
  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
}

const scenarios = [
  { alg: 'md5', digest: 'hex' },
  { alg: 'md5', digest: 'base64' },
  { alg: 'sha1', digest: 'hex' },
  { alg: 'sha1', digest: 'base64' },
  { alg: 'sha256', digest: 'hex' },
  { alg: 'sha256', digest: 'base64' },
  { alg: 'cyrb53', digest: 'hex' }
]

for (const { alg, digest } of scenarios) {
    suite.add(`${alg}-${digest}`, () => {
        if (alg === 'cyrb53') cyrb53(data)
        else hash(alg).update(data).digest(digest)
    }
  )
}
suite.on('cycle', function (event: { target: string }) {
  console.log(String(event.target))
})
  .on('complete', function (this: any) {
    console.log('Fastest is ' + this.filter('fastest').map('name'))
  })
  .run()
```

</details>

ts-node is much faster than bun on all `node:crypto` algorithms, except for the pure JS hashing algorithm on my machine.

```
/home/u9g/code/2-25-23/show-watcher〉ts-node suite.ts
md5-hex x 589,929 ops/sec ±3.80% (79 runs sampled)
md5-base64 x 569,748 ops/sec ±4.07% (81 runs sampled)
sha1-hex x 625,474 ops/sec ±3.69% (76 runs sampled)
sha1-base64 x 609,722 ops/sec ±3.76% (75 runs sampled)
sha256-hex x 504,345 ops/sec ±3.04% (83 runs sampled)
sha256-base64 x 501,889 ops/sec ±3.26% (86 runs sampled)
cyrb53-hex x 1,934,808 ops/sec ±1.01% (93 runs sampled)
Fastest is cyrb53-hex
/home/u9g/code/2-25-23/show-watcher〉bun run suite.ts
md5-hex x 486,613 ops/sec ±0.89% (93 runs sampled)
md5-base64 x 468,318 ops/sec ±1.72% (92 runs sampled)
sha1-hex x 467,392 ops/sec ±1.26% (93 runs sampled)
sha1-base64 x 463,872 ops/sec ±1.53% (90 runs sampled)
sha256-hex x 384,813 ops/sec ±1.04% (95 runs sampled)
sha256-base64 x 391,336 ops/sec ±1.01% (92 runs sampled)
cyrb53-hex x 2,787,723 ops/sec ±0.33% (95 runs sampled)
Fastest is cyrb53-hex
```

### PR DESCRIPTION
**Pull request by paperclover**
Currently, `node:events` is implemented in C++, which may sound fast but ends up being much slower than a pure JS implementation as the rapid hopping between native code and JS with listener callbacks is a lot of overhead. This PR is not only porting the node.js `events` into JS as closely as possible, but also trying to get it as fast as possible.

Currently node.js own events implementation beats us, but we've gone from being 4x slower to something more like 2x.

Still debating which implementation details are best, such as
- `Object` vs `Map`: they lead to different gains in different tests
- An optimization where having only one listener is not an array but just the function itself. This seems to be slower than just having and looping over a length=1 array.

Also, setting real-world benchmarks that tests actual usage instead of than a single ".emit()". Microbenchmarks are handy but we need to consider the bigger picture.

**required todo**:

- [x] Replace the C++ module out for `events.exports.js`
- [x] Find the best implementations for `addListener` and `emit`
  - the array clone we have to do seems to be the slow part?
- [x] `EventEmitter#once` (implemented but check if can be made faster)
  - [x] optimization: compare arrow fn vs the current bind solution.
- [x] `EventEmitter#removeListener` (reimplement)
- [x] `EventEmitter#prependListener`
- [x] `EventEmitter#listenerCount`
- [x] `EventEmitter#types`
- [x] `EventEmitter.once` (async iterator)
- [x] `EventEmitter.setMaxListeners`
- [x] `EventEmitter.getEventListeners`
- [x] error event
- [x] capturing rejections
- [x] validators

**things might not do in this PR**

- [ ] Remove the C++ implementation, it's still in the code somewhere.
- [ ] `EventEmitter.on` (async iterator)
- [ ] `EventEmitterAsyncResource`

Fixes #2190

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
a little faster with this:
```js
// Reimplementation of https://nodejs.org/api/events.html
// Reference: https://github.com/nodejs/node/blob/main/lib/events.js
const SymbolFor = Symbol.for;
const ObjectDefineProperty = Object.defineProperty;
const kCapture = Symbol("kCapture");
const kErrorMonitor = Symbol("events.errorMonitor");
const kMaxEventTargetListeners = Symbol("events.maxEventTargetListeners");
const kMaxEventTargetListenersWarned = Symbol("events.maxEventTargetListenersWarned");
const kWatermarkData = Symbol.for("nodejs.watermarkData");
const kRejection = SymbolFor("nodejs.rejection");
const captureRejectionSymbol = Symbol.for("nodejs.rejection");
const ArrayPrototypeSlice = Array.prototype.slice;
var { isPromise } = import.meta.primordials;

var errorMonitor = Symbol.for("events.errorMonitor");

var defaultMaxListeners = 10;

// EventEmitter must be a standard function because some old code will do weird tricks like `EventEmitter.apply(this)`.
function EventEmitter(opts) {
  if (this._events === undefined || this._events === this.__proto__._events) {
    this._events = { __proto__: null };
    this._eventsCount = 0;
  }

  this._maxListeners ??= undefined;
  if (
    (this[kCapture] = opts?.captureRejections ? Boolean(opts?.captureRejections) : EventEmitter.prototype[kCapture])
  ) {
    this.emit = emitWithRejectionCapture;
  }
}

EventEmitter.prototype._events = undefined;
EventEmitter.prototype._eventsCount = 0;
EventEmitter.prototype._maxListeners = undefined;
EventEmitter.prototype.setMaxListeners = function setMaxListeners(n) {
  validateNumber(n, "setMaxListeners", 0);
  this._maxListeners = n;
  return this;
};

EventEmitter.prototype.getMaxListeners = function getMaxListeners() {
  return this._maxListeners ?? EventEmitter.defaultMaxListeners;
};

function emitError(emitter, args) {
  var { _events: events } = emitter;
  if (!events) throw args[0];
  var handlers = events.error;
  if (!handlers) throw args[0];
  for (var handler of handlers.slice()) {
    handler.apply(emitter, args);
  }
  return true;
}

function addCatch(emitter, promise, type, args) {
  if (!emitter[kCapture]) {
    return;
  }
  // Handle Promises/A+ spec, then could be a getter that throws on second use.
  try {
    const then = promise.then;
    if (typeof then === "function") {
      then.call(promise, undefined, function (err) {
        // The callback is called with queueMicrotask to avoid a follow-up rejection from this promise.
        queueMicrotask(emitUnhandledRejectionOrErr, that, err, type, args);
      });
    }
  } catch (err) {
    that.emit("error", err);
  }
}

function emitUnhandledRejectionOrErr(emitter, err, type, args) {
  if (typeof emitter[kRejection] === "function") {
    emitter[kRejection](err, type, ...args);
  } else {
    // If the error handler throws, it is not catchable and it will end up in 'uncaughtException'.
    // We restore the previous value of kCapture in case the uncaughtException is present
    // and the exception is handled.
    try {
      emitter[kCapture] = false;
      emitter.emit("error", err);
    } finally {
      emitter[kCapture] = true;
    }
  }
}

const emitWithoutRejectionCapture = function emit(type, ...args) {
  if (type === "error") {
    return emitError(this, args);
  }
  var { _events: events } = this;
  if (events === undefined) return false;
  var handlers = events[type];
  if (handlers === undefined) return false;

  for (var handler of [...handlers]) {
    handler.apply(this, args);
  }
  return true;
};

const emitWithRejectionCapture = function emit(type, ...args) {
  if (type === "error") {
    return emitError(this, args);
  }
  var { _events: events } = this;
  if (events === undefined) return false;
  var handlers = events[type];
  if (handlers === undefined) return false;
  for (var handler of handlers.slice()) {
    var result = handler.apply(this, args);
    if (result !== undefined && isPromise(result)) {
      addCatch(this, result, type, args);
    }
  }
  return true;
};

EventEmitter.prototype.emit = emitWithoutRejectionCapture;

EventEmitter.prototype.addListener = function addListener(type, fn) {
  var events = this._events;
  if (!events) {
    events = this._events = { __proto__: null };
    this._eventsCount = 0;
  } else if (events.newListener) {
    this.emit("newListener", type, fn.listener ?? fn);
  }
  var handlers = events[type];
  if (!handlers) {
    events[type] = [fn];
    this._eventsCount++;
  } else {
    handlers.push(fn);
    // TODO: overflow check
  }
  return this;
};

EventEmitter.prototype.on = EventEmitter.prototype.addListener;

EventEmitter.prototype.prependListener = function prependListener(type, fn) {
  var events = this._events;
  if (!events) {
    events = this._events = { __proto__: null };
    this._eventsCount = 0;
  } else if (events.newListener) {
    this.emit("newListener", type, fn.listener ?? fn);
  }
  var handlers = events[type];
  if (!handlers) {
    events[type] = [fn];
    this._eventsCount++;
  } else {
    handlers.unshift(fn);
    // TODO: overflow check
  }
  return this;
};

function onceWrapper(type, listener, ...args) {
  this.removeListener(type, listener);
  listener.apply(this, args);
}

EventEmitter.prototype.once = function once(type, listener) {
  const bound = onceWrapper.bind(this, type, listener);
  bound.listener = listener;
  this.addListener(type, bound);
  return this;
};

EventEmitter.prototype.prependOnceListener = function prependOnceListener(type, fn) {
  const bound = onceWrapper.bind(this, type, fn);
  bound.listener = fn;
  this.prependListener(type, bound);
  return this;
};

EventEmitter.prototype.removeListener = function removeListener(type, fn) {
  var { _events: events } = this;
  if (!events) return this;
  var handlers = events[type];
  if (!handlers) return this;
  var length = handlers.length;
  let position = -1;
  for (let i = length - 1; i >= 0; i--) {
    if (handlers[i] === fn || handlers[i].listener === fn) {
      position = i;
      break;
    }
  }
  if (position < 0) return this;
  if (position === 0) {
    handlers.shift();
  } else {
    handlers.splice(position, 1);
  }
  if (handlers.length === 0) {
    delete events[type];
    this._eventsCount--;
  }
  return this;
};

EventEmitter.prototype.off = EventEmitter.prototype.removeListener;

EventEmitter.prototype.removeAllListeners = function removeAllListeners(type) {
  var { _events: events } = this;
  if (type && events) {
    if (events[type]) {
      delete events[type];
      this._eventsCount--;
    }
  } else {
    this._events = { __proto__: null };
  }
  return this;
};

EventEmitter.prototype.listeners = function listeners(type) {
  var { _events: events } = this;
  if (!events) return [];
  var handlers = events[type];
  if (!handlers) return [];
  return handlers.map(x => x.listener ?? x);
};

EventEmitter.prototype.rawListeners = function rawListeners(type) {
  var { _events } = this;
  if (!_events) return [];
  var handlers = _events[type];
  if (!handlers) return [];
  return handlers.slice();
};

EventEmitter.prototype.listenerCount = function listenerCount(type) {
  var { _events: events } = this;
  if (!events) return 0;
  return events[type]?.length ?? 0;
};

EventEmitter.prototype.eventNames = function eventNames() {
  return this._eventsCount > 0 ? Reflect.ownKeys(this._events) : [];
};

EventEmitter.prototype[kCapture] = false;

function once(emitter, type, { signal } = {}) {
  validateAbortSignal(signal, "options.signal");
  if (signal?.aborted) {
    // TODO: use AbortError
    throw new Error(undefined, { cause: signal?.reason });
  }
  return new Promise((resolve, reject) => {
    const errorListener = err => {
      emitter.removeListener(type, resolver);
      if (signal != null) {
        eventTargetAgnosticRemoveListener(signal, "abort", abortListener);
      }
      reject(err);
    };
    const resolver = (...args) => {
      if (typeof emitter.removeListener === "function") {
        emitter.removeListener("error", errorListener);
      }
      if (signal != null) {
        eventTargetAgnosticRemoveListener(signal, "abort", abortListener);
      }
      resolve(args);
    };
    eventTargetAgnosticAddListener(emitter, type, resolver, { once: true });
    if (type !== "error" && typeof emitter.once === "function") {
      // EventTarget does not have `error` event semantics like Node
      // EventEmitters, we listen to `error` events only on EventEmitters.
      emitter.once("error", errorListener);
    }
    function abortListener() {
      eventTargetAgnosticRemoveListener(emitter, type, resolver);
      eventTargetAgnosticRemoveListener(emitter, "error", errorListener);
      // TODO: use AbortError
      reject(new Error(undefined, { cause: signal?.reason }));
    }
    if (signal != null) {
      eventTargetAgnosticAddListener(signal, "abort", abortListener, { once: true });
    }
  });
}
EventEmitter.once = once;

function on(emitter, type, { signal, close, highWatermark = Number.MAX_SAFE_INTEGER, lowWatermark = 1 } = {}) {
  throw new Error("events.on is not implemented");
}
EventEmitter.on = on;

function getEventListeners(emitter, type) {
  // TODO: EventTarget support
  return emitter.listeners(type);
}
EventEmitter.getEventListeners = getEventListeners;

function setMaxListeners(n, ...eventTargets) {
  validateNumber(n, "setMaxListeners", 0);
  if (eventTargets) {
    var { length } = eventTargets;
    for (let i = 0; i < length; i++) {
      eventTargets[i].setMaxListeners(n);
    }
  } else {
    defaultMaxListeners = n;
  }
}
EventEmitter.setMaxListeners = setMaxListeners;

function listenerCount(emitter, type) {
  return emitter.listenerCount(type);
}
EventEmitter.listenerCount = listenerCount;

EventEmitter.EventEmitter = EventEmitter;
EventEmitter.usingDomains = false;
EventEmitter.captureRejectionSymbol = captureRejectionSymbol;
ObjectDefineProperty(EventEmitter, "captureRejections", {
  __proto__: null,
  get() {
    return EventEmitter.prototype[kCapture];
  },
  set(value) {
    validateBoolean(value, "EventEmitter.captureRejections");

    EventEmitter.prototype[kCapture] = value;
  },
  enumerable: true,
});
EventEmitter.errorMonitor = errorMonitor;
Object.defineProperties(EventEmitter, {
  defaultMaxListeners: {
    enumerable: true,
    get: () => {
      return defaultMaxListeners;
    },
    set: arg => {
      validateNumber(arg, "defaultMaxListeners", 0);
      defaultMaxListeners = arg;
    },
  },
  kMaxEventTargetListeners: {
    __proto__: null,
    value: kMaxEventTargetListeners,
    enumerable: false,
    configurable: false,
    writable: false,
  },
  kMaxEventTargetListenersWarned: {
    __proto__: null,
    value: kMaxEventTargetListenersWarned,
    enumerable: false,
    configurable: false,
    writable: false,
  },
});
EventEmitter.init = EventEmitter;
EventEmitter[Symbol.for("CommonJS")] = 0;

export default EventEmitter;

function eventTargetAgnosticRemoveListener(emitter, name, listener, flags) {
  if (typeof emitter.removeListener === "function") {
    emitter.removeListener(name, listener);
  } else {
    emitter.removeEventListener(name, listener, flags);
  }
}

function eventTargetAgnosticAddListener(emitter, name, listener, flags) {
  if (typeof emitter.on === "function") {
    emitter.on(name, listener);
  } else {
    emitter.addEventListener(name, listener);
  }
}

function ERR_INVALID_ARG_TYPE(name, type, value) {
  const err = new TypeError(`The "${name}" argument must be of type ${type}. Received ${value}`);
  err.code = "ERR_INVALID_ARG_TYPE";
  return err;
}

function ERR_OUT_OF_RANGE(name, range, value) {
  const err = new RangeError(`The "${name}" argument is out of range. It must be ${range}. Received ${value}`);
  err.code = "ERR_OUT_OF_RANGE";
  return err;
}

function validateAbortSignal(signal, name) {
  if (signal !== undefined && (signal === null || typeof signal !== "object" || !("aborted" in signal))) {
    throw new ERR_INVALID_ARG_TYPE(name, "AbortSignal", signal);
  }
}

function validateNumber(value, name, min = undefined, max) {
  if (typeof value !== "number") throw new ERR_INVALID_ARG_TYPE(name, "number", value);
  if (
    (min != null && value < min) ||
    (max != null && value > max) ||
    ((min != null || max != null) && Number.isNaN(value))
  ) {
    throw new ERR_OUT_OF_RANGE(
      name,
      `${min != null ? `>= ${min}` : ""}${min != null && max != null ? " && " : ""}${max != null ? `<= ${max}` : ""}`,
      value,
    );
  }
}

export class EventEmitterAsyncResource extends EventEmitter {
  constructor(options = undefined) {
    throw new Error("EventEmitterAsyncResource is not implemented");
  }
}

EventEmitter.EventEmitterAsyncResource = EventEmitterAsyncResource;
```

**Comment by paperclover** [comment]
3 opened issues for less important but missing features. passing all tests for everything else though

**Comment by Jarred-Sumner** [comment]
looks like there are unintentional changes to submodules

can you reset those?

File: test/package.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "esbuild": "^0.17.16",
     "express": "^4.18.2",
     "iconv-lite": "^0.6.3",
+    "jest": "^29.5.0",
```
Comment (Jarred-Sumner) [inline test/package.json]: is this intentional?

File: test/js/node/events/event-emitter.test.ts [hunk 2]
```diff
@@ -1,34 +1,106 @@
...
+  test("once (abort)", done => {
+    const emitter = new EventEmitter();
+    const controller = new AbortController();
+    EventEmitter.once(emitter, "hey", { signal: controller.signal })
+      .then(() => done(new Error("Should not be called")))
+      .catch(() => done());
+    controller.abort();
+  });
+
+  test("once (two events in same tick)", done => {
```
Comment (Jarred-Sumner) [inline test/js/node/events/event-emitter.test.ts:37]: nice

File: src/bun.js/events.exports.js [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+const ObjectDefineProperty = Object.defineProperty;
+const kCapture = Symbol("kCapture");
+const kErrorMonitor = SymbolFor("events.errorMonitor");
+const kMaxEventTargetListeners = Symbol("events.maxEventTargetListeners");
+const kMaxEventTargetListenersWarned = Symbol("events.maxEventTargetListenersWarned");
+const kWatermarkData = SymbolFor("nodejs.watermarkData");
+const kRejection = SymbolFor("nodejs.rejection");
+const captureRejectionSymbol = SymbolFor("nodejs.rejection");
+const ArrayPrototypeSlice = Array.prototype.slice;
+var { isPromise } = import.meta.primordials;
```
Comment (Jarred-Sumner) [inline src/bun.js/events.exports.js]: We can make `Array` and `Object` come from primordials here

```suggestion
var { isPromise, Array, Object } = import.meta.primordials;
```

File: test/package.json (hunk 1 above)
Comment (paperclover) [reply test/package.json]: accidentally added when i was trying to run the node event emitter test with jest. good catch.

**Pull request merged**